File: src/sunday/system/protocols.py
"""Structural protocols for substituting fakes in place of JarvisSystem."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, List, Optional, Protocol

if TYPE_CHECKING:
    from sunday.core.config import JarvisConfig
    from sunday.core.events import EventBus
    from sunday.engine._stubs import InferenceEngine
    from sunday.security.capabilities import CapabilityPolicy
    from sunday.sessions.session import SessionStore
    from sunday.tools._stubs import BaseTool
    from sunday.tools.storage._stubs import MemoryBackend
    from sunday.traces.collector import TraceCollector
    from sunday.traces.store import TraceStore


class OrchestratorDeps(Protocol):
    """Minimum surface of JarvisSystem that QueryOrchestrator depends on.

    Tests can satisfy this with a lightweight class — no need to construct
    the full JarvisSystem dataclass or materialize every subsystem.
    """

    config: JarvisConfig
    bus: EventBus
    engine: InferenceEngine
    engine_key: str
    model: str
    agent_name: str
    tools: List[BaseTool]
    memory_backend: Optional[MemoryBackend]
    capability_policy: Optional[CapabilityPolicy]
    session_store: Optional[SessionStore]
    trace_store: Optional[TraceStore]
    trace_collector: Optional[TraceCollector]  # written by _run_agent

    # Optional attribute (getattr with default) — declared for type clarity.
    _skill_few_shot_examples: Any

File: rust/crates/sunday-system/src/protocols.rs
//! Structural protocols for substituting fakes in place of JarvisSystem.
//!
//! Rust translation of `src/sunday/system/protocols.py`.

use sunday_core::{EventBus, JarvisConfig};
use sunday_engine::InferenceEngine;
use std::sync::Arc;

/// Minimum surface of JarvisSystem that QueryOrchestrator depends on.
///
/// Tests can satisfy this with a lightweight struct — no need to construct
/// the full JarvisSystem or materialize every subsystem.
pub trait OrchestratorDeps: Send + Sync {
    fn config(&self) -> &JarvisConfig;
    fn bus(&self) -> &EventBus;
    fn engine(&self) -> &dyn InferenceEngine;
    fn engine_key(&self) -> &str;
    fn model(&self) -> &str;
    fn agent_name(&self) -> &str;
    fn memory_backend(&self) -> Option<&Arc<dyn Send + Sync>>;
    fn capability_policy(&self) -> Option<&Arc<dyn Send + Sync>>;
    fn session_store(&self) -> Option<&Arc<dyn Send + Sync>>;
    fn trace_store(&self) -> Option<&Arc<dyn Send + Sync>>;
    fn trace_collector(&self) -> Option<&Arc<dyn Send + Sync>>;
}
